rust_src/src/windows.rs

```diff
@@ -1536,4 +1592,132 @@ pub fn window_body_width_lisp(window: LispWindowLiveOrSelected, pixelwise: bool)
+/// (body edge if BODY is non-nil) of WINDOW.  If LEFT and INVERSE are both
+/// non-nil, the y-pixel value returned for a specific line specifies the
+/// distance in pixels from the left edge (body edge if BODY is non-nil) of
+/// WINDOW to the left edge of the last (leftmost) glyph of that line.
+///
+/// Normally, the value of this function is not available while Emacs is
+/// busy, for example, when processing a command.  It should be retrievable
+/// though when run from an idle timer with a delay of zero seconds.
+#[lisp_fn(min = "0")]
+pub fn window_lines_pixel_dimensions(
+    window: LispWindowLiveOrSelected,
+    first: Option<LispNumber>,
+    last: Option<LispNumber>,
+    body: bool,
+    inverse: bool,
+    left: bool,
+) -> LispObject {
+    let window: LispWindowRef = window.into();
+    let buffer: LispBufferRef = window.contents.into();
+    let max_y = if body {
+        window.text_bottom_y()
+    } else {
+        window.pixel_height
+    };
+    let window_width = if body {
+        window.body_width(true)
+    } else {
+        window.pixel_width
+    };
+    let header_line_height = window.header_line_height();
+    let subtract = if body { header_line_height } else { 0 };
+    let matrix = window.get_matrix();
+
+    if unsafe { noninteractive } || window.is_pseudo() {
+        return Qnil;
+    }
+
+    // Fail if current matrix is not up to date.
+    unsafe {
+        if !window.window_end_valid()
+            || windows_or_buffers_changed > 0
+            || buffer.clip_changed()
+            || buffer.prevent_redisplay_optimizations_p()
+            || window.is_outdated()
+        {
+            return Qnil;
+        }
+    }
+
+    let mut row = match first {
+        Some(first) => matrix.row(first.to_fixnum() as usize),
+        None => {
+            if body {
+                matrix.first_text_row()
+            } else {
+                matrix.row(0)
+            }
+        }
+    };
+
+    let end_row = match last {
+        Some(last) => matrix.row(last.to_fixnum() as usize),
+        None => {
+            if body {
+                matrix.bottom_text_row(window)
+            } else {
+                unsafe { matrix.row_unchecked(matrix.nrows as usize) }
+            }
+        }
+    };
+
+    let mut rows = Qnil;
+    while row.as_ptr() <= end_row.as_ptr() && row.enabled_p() && row.y + row.height < max_y {
+        let width = if left {
+            let glyph = unsafe { &*row.glyphs[TEXT_AREA as usize] };
+            if inverse {
+                glyph.pixel_width as i32
+            } else {
+                window_width - glyph.pixel_width as i32
+            }
+        } else {
+            if inverse {
+                window_width - row.pixel_width
+            } else {
+                row.pixel_width
+            }
+        };
+        rows = ((width, row.y + row.height - subtract), rows).into();
+        let ptr = unsafe { row.as_mut().add(1) };
```

Comment:
The incrementing of the contained pointer feels like functionality the `ExternalPtr` type should make easier. These two lines could be something like `row.ptr_add(1)`. Maybe a follow on PR?

Reply:
I opened #1317